From 45bc413b2b87aaaab7a2cc02d2cd018edc7531ee Mon Sep 17 00:00:00 2001
From: Justin Maggard <jmaggard@netgear.com>
Date: Wed, 9 Nov 2016 14:43:17 -0800
Subject: [PATCH 1/2] fix guest mounting

When doing a guest mount, fill in a bogus username if none was
specified. Windows 10 at least requires *something* in the username
field. If that fails, then retry with the default behavior of not
sending a username.
---
 mount.cifs.c | 32 +++++++++++++++++++++-----------
 1 file changed, 21 insertions(+), 11 deletions(-)

diff --git a/mount.cifs.c b/mount.cifs.c
index 5c5734f..0569911 100644
--- a/mount.cifs.c
+++ b/mount.cifs.c
@@ -1053,6 +1053,12 @@ parse_options(const char *data, struct parsed_mount_info *parsed_info)
 			*filesys_flags &= ~MS_NOEXEC;
 			goto nocopy;
 		case OPT_GUEST:
+			if (!parsed_info->got_user)
+			{
+				strlcpy(txtbuf, "GuestXXXXXX", sizeof(txtbuf));
+				strlcpy(parsed_info->username, mktemp(txtbuf),
+					sizeof(parsed_info->username));
+			}
 			parsed_info->got_user = 1;
 			parsed_info->got_password = 1;
 			goto nocopy;
@@ -1793,17 +1799,6 @@ assemble_mountinfo(struct parsed_mount_info *parsed_info,
 			goto assemble_exit;
 	}
 
-	/* copy in user= string */
-	if (parsed_info->got_user) {
-		if (*parsed_info->options)
-			strlcat(parsed_info->options, ",",
-				sizeof(parsed_info->options));
-		strlcat(parsed_info->options, "user=",
-			sizeof(parsed_info->options));
-		strlcat(parsed_info->options, parsed_info->username,
-			sizeof(parsed_info->options));
-	}
-
 	if (*parsed_info->domain) {
 		if (*parsed_info->options)
 			strlcat(parsed_info->options, ",",
@@ -2056,6 +2051,11 @@ mount_retry:
 		fprintf(stderr, "%s kernel mount options: %s",
 			thisprogram, options);
 
+	if (parsed_info->got_user) {
+		strlcat(options, ",user=", options_size);
+		strlcat(options, parsed_info->username, options_size);
+	}
+
 	if (parsed_info->got_password) {
 		/*
 		 * Commas have to be doubled, or else they will
@@ -2082,6 +2082,16 @@ mount_retry:
 			goto do_mtab;
 
 		switch (errno) {
+		case EACCES:
+			if (parsed_info->got_password && !parsed_info->password[0] &&
+			    parsed_info->got_user &&
+			    !strncmp(parsed_info->username, "Guest", 5)) {
+				fprintf(stderr,
+					"Retrying guest mount with no user name\n");
+				parsed_info->username[0] = '\0';
+				goto mount_retry;
+			}
+			break;
 		case ECONNREFUSED:
 		case EHOSTUNREACH:
 			currentaddress = nextaddress;
-- 
2.10.2

